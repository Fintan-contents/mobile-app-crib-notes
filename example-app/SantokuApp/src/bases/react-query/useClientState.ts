import {useCallback, useMemo, useRef} from 'react';
import type {QueryKey} from 'react-query';
import {hashQueryKey, useQuery, useQueryClient} from 'react-query';
export type ClientStateKey = string | number | QueryKey;
export type InitialClientState<S> = S | (() => S);
export type SetClientStateFunction<S> = (prevState: S | undefined) => S;
export type SetClientState<S> = (value: S | SetClientStateFunction<S>) => void;
export type ClientStateAndSetClientState<S> = [S, SetClientState<S>];
export type ClientStateOptions = {
  preserveAfterUnmount?: boolean;
};
export function useClientState<S = undefined>(
  key: ClientStateKey,
  options?: ClientStateOptions,
): ClientStateAndSetClientState<S | undefined>;
export function useClientState<S>(
  key: ClientStateKey,
  initialState: InitialClientState<S>,
  options?: ClientStateOptions,
): ClientStateAndSetClientState<S>;
export function useClientState<S>(
  key: ClientStateKey,
  initialState?: InitialClientState<S | undefined>,
  // preserveAfterUnmount: boolean = false,
  options: ClientStateOptions = {preserveAfterUnmount: false},
): ClientStateAndSetClientState<S | undefined> {
  const queryKey = useStableQueryKey(key);
  const data = useQuery<S | null>(queryKey, {
    enabled: false,
    initialData: initialState,
    cacheTime: options.preserveAfterUnmount ? Number.POSITIVE_INFINITY : undefined,
  }).data;
  const state = data === null ? undefined : data;
  const setState = useSetClientStateFunction<S | undefined>(queryKey);
  return [state, setState];
}
export function useOnlySetClientState<S>(key: ClientStateKey) {
  const queryKey = useStableQueryKey(key);
  return useSetClientStateFunction<S>(queryKey);
}
function useStableQueryKey(key: ClientStateKey): unknown[] {
  const queryKeyRef = useRef(ensureArrayKey(key));
  const queryKeyHashRef = useRef(hashQueryKey(queryKeyRef.current));
  return useMemo(() => {
    const newQueryKey = ensureArrayKey(key);
    const newQueryKeyHash = hashQueryKey(newQueryKey);
    if (queryKeyHashRef.current === newQueryKeyHash) {
      return queryKeyRef.current;
    }
    queryKeyRef.current = newQueryKey;
    queryKeyHashRef.current = newQueryKeyHash;
    return newQueryKey;
  }, [key]);
}
function useSetClientStateFunction<S>(queryKey: QueryKey): SetClientState<S> {
  const queryClient = useQueryClient();
  return useCallback(
    (value: S | SetClientStateFunction<S>) => {
      const newValue = isSetClientStateFunction(value) ? value(queryClient.getQueryData(queryKey)) : value;
      queryClient.setQueryData<S | null>(queryKey, newValue === undefined ? null : newValue);
    },
    [queryClient, queryKey],
  );
}
function isSetClientStateFunction<S>(arg: S | SetClientStateFunction<S>): arg is SetClientStateFunction<S> {
  return typeof arg === 'function';
}
function ensureArrayKey(key: ClientStateKey): unknown[] {
  return Array.isArray(key) ? key : [key];
}
