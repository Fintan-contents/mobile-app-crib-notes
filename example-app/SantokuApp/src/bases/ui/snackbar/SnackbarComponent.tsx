import React, {useCallback, useRef, useState} from 'react';
import {Animated, StyleProp, StyleSheet, Text, TextStyle, TouchableOpacity, View, ViewStyle} from 'react-native';

import CompositeAnimation = Animated.CompositeAnimation;

export type SnackbarShowProps = {
  /**
   * Displayed message.
   */
  message: string;
  /**
   * Style of message.
   */
  messageTextStyle?: StyleProp<TextStyle>;
  /**
   * Style to adjust the look of the snackbar.
   */
  style?: StyleProp<ViewStyle>;
  /**
   * Adjust the position of the snackbar style.
   */
  positionStyle?: StyleProp<ViewStyle>;
  /**
   * Button text.
   * Placed to the right of the message.
   */
  actionText?: string;
  /**
   * Function called when the user taps the button.
   */
  actionHandler?: () => void;
  /**
   * Style of Button text.
   */
  actionTextStyle?: StyleProp<TextStyle>;
  /**
   * Time from when the snackbar appears to when it starts to fade-out.
   */
  autoHideDuration?: number;
  /**
   * Snackbar fade-in time.
   */
  fadeInDuration?: number;
  /**
   * Snackbar fade-out time.
   */
  fadeOutDuration?: number;
  /**
   * Fade-out time to close the previously displayed snack bar when displaying a new one.
   * This is applied when you try to display another snackbar while a snackbar is being displayed.
   */
  forceFadeOutDuration?: number;
  /**
   * Timestamp
   */
  timestamp?: number;
};

export type SnackbarHideProps = {
  /**
   * Specify when to hide the snackbar.
   */
  hide?: true;
  /**
   * Fade-out time when hide is specified.
   */
  hideFadeOutDuration?: number;
};

export type SnackbarProps = SnackbarShowProps & SnackbarHideProps;

export const SnackbarComponent: React.FC<SnackbarProps> = props => {
  const fadeAnim = useRef(new Animated.Value(0)).current;
  const fadeInAnimationRef = useRef<CompositeAnimation>();
  const fadeOutAnimationRef = useRef<CompositeAnimation>();
  const barrierFadeOutAnimationRef = useRef<CompositeAnimation>();
  const [visibleSnackbarProps, setVisibleSnackbarProps] = useState<SnackbarShowProps>();

  const animationStart = useCallback(
    (
      animationRef: React.MutableRefObject<CompositeAnimation | undefined>,
      config: Animated.TimingAnimationConfig,
      callback?: Animated.EndCallback,
    ) => {
      const animation = Animated.timing(fadeAnim, config);
      animationRef.current = animation;
      animation.start(result => {
        animationRef.current = undefined;
        if (callback) {
          callback(result);
        }
      });
    },
    [fadeAnim],
  );

  const show = useCallback(() => {
    setVisibleSnackbarProps({
      actionHandler: props.actionHandler,
      actionText: props.actionText,
      actionTextStyle: props.actionTextStyle,
      autoHideDuration: props.autoHideDuration,
      fadeInDuration: props.fadeInDuration,
      fadeOutDuration: props.fadeOutDuration,
      forceFadeOutDuration: props.forceFadeOutDuration,
      message: props.message,
      messageTextStyle: props.messageTextStyle,
      positionStyle: props.positionStyle,
      style: props.style,
      timestamp: props.timestamp,
    });

    const fadeInAnimationConfig = {
      toValue: 1,
      duration: props.fadeInDuration,
      useNativeDriver: true,
    };
    animationStart(fadeInAnimationRef, fadeInAnimationConfig, ({finished}) => {
      if (finished) {
        const fadeOutAnimationConfig = {
          toValue: 0,
          delay: props.autoHideDuration,
          duration: props.fadeOutDuration,
          useNativeDriver: true,
        };
        animationStart(fadeOutAnimationRef, fadeOutAnimationConfig, () => {
          if (!barrierFadeOutAnimationRef.current) {
            setVisibleSnackbarProps(undefined);
          }
        });
      } else {
        if (!barrierFadeOutAnimationRef.current) {
          setVisibleSnackbarProps(undefined);
        }
      }
    });
  }, [
    animationStart,
    props.actionHandler,
    props.actionText,
    props.actionTextStyle,
    props.autoHideDuration,
    props.fadeInDuration,
    props.fadeOutDuration,
    props.forceFadeOutDuration,
    props.message,
    props.messageTextStyle,
    props.positionStyle,
    props.style,
    props.timestamp,
  ]);

  const forceFadeout = useCallback(
    (fadeOutDuration?: number, callback?: () => void) => {
      const barrierFadeOutAnimationConfig = {
        toValue: 0,
        duration: fadeOutDuration,
        useNativeDriver: true,
      };

      barrierFadeOutAnimationRef.current?.stop();

      animationStart(barrierFadeOutAnimationRef, barrierFadeOutAnimationConfig, ({finished}) => {
        if (finished) {
          setVisibleSnackbarProps(undefined);
          callback?.();
        }
      });
      fadeInAnimationRef.current?.stop();
      fadeOutAnimationRef.current?.stop();
    },
    [animationStart],
  );

  React.useEffect(() => {
    if (props.hide) {
      forceFadeout(props.hideFadeOutDuration);
      return;
    }
    if (!props.message) {
      return;
    }
    if (barrierFadeOutAnimationRef.current) {
      forceFadeout(props.forceFadeOutDuration, show);
      return;
    }
    if (fadeInAnimationRef.current || fadeOutAnimationRef.current) {
      forceFadeout(props.forceFadeOutDuration, show);
      return;
    }
    show();
  }, [show, forceFadeout, props.hide, props.message, props.hideFadeOutDuration, props.forceFadeOutDuration]);

  const snackbarStyle = StyleSheet.flatten([styles.snackbar, props.style]);

  const animatedViewStyle = StyleSheet.flatten<ViewStyle>([styles.animatedContainer, props.positionStyle]);
  if (animatedViewStyle?.top === undefined && animatedViewStyle.bottom === undefined) {
    animatedViewStyle.bottom = 20;
  }

  const messageTextStyle = StyleSheet.flatten([styles.messageText, props.messageTextStyle]);
  const actionTextStyle = StyleSheet.flatten([styles.actionText, props.actionTextStyle]);

  if (!visibleSnackbarProps) {
    return null;
  }

  return (
    <Animated.View style={StyleSheet.flatten([{opacity: fadeAnim}, animatedViewStyle])} testID="snackbarAnimatedView">
      <View style={snackbarStyle} testID="snackbarStyleView">
        <View style={styles.messageContainer}>
          <Text style={messageTextStyle}>{visibleSnackbarProps.message}</Text>
        </View>
        {visibleSnackbarProps.actionText && visibleSnackbarProps.actionHandler && (
          <View style={styles.actionContainer}>
            <TouchableOpacity onPress={visibleSnackbarProps.actionHandler}>
              <Text style={actionTextStyle}>{visibleSnackbarProps.actionText}</Text>
            </TouchableOpacity>
          </View>
        )}
      </View>
    </Animated.View>
  );
};

SnackbarComponent.defaultProps = {
  autoHideDuration: 4000,
  fadeInDuration: 1000,
  fadeOutDuration: 1000,
  forceFadeOutDuration: 300,
  hideFadeOutDuration: 300,
};

const styles = StyleSheet.create({
  animatedContainer: {
    flex: 1,
    position: 'absolute',
    minHeight: 48,
    paddingHorizontal: 10,
    width: '100%',
  },
  snackbar: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 5,
    backgroundColor: '#393939',
  },
  messageContainer: {
    flex: 1,
    paddingVertical: 20,
    paddingLeft: 20,
    paddingRight: 10,
  },
  actionContainer: {
    flex: 0,
    justifyContent: 'flex-end',
    alignItems: 'center',
    paddingVertical: 20,
    paddingLeft: 10,
    paddingRight: 20,
  },
  messageText: {
    color: 'white',
  },
  actionText: {
    color: '#85A3F1',
  },
});
