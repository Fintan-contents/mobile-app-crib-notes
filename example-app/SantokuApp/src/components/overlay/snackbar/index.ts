export * from './Snackbar';
export * from './WithSnackbar';
