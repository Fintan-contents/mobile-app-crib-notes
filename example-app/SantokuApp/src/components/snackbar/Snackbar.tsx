import React, {useCallback, useRef, useState} from 'react';
import {Animated, StyleProp, StyleSheet, Text, TextStyle, TouchableOpacity, View, ViewStyle} from 'react-native';

import {FullWindowOverlay} from './FullWindowOverlay';

import CompositeAnimation = Animated.CompositeAnimation;

export type SnackbarShowProps = {
  message: string;
  messageTextStyle?: StyleProp<TextStyle>;
  top?: number;
  bottom?: number;
  right?: number;
  left?: number;
  style?: StyleProp<ViewStyle>;
  animatedContainerStyle?: StyleProp<ViewStyle>;
  actionText?: string;
  actionHandler?: () => void;
  actionTextStyle?: StyleProp<TextStyle>;
  autoHideDuration?: number;
  fadeInDuration?: number;
  fadeOutDuration?: number;
  forceFadeOutDuration?: number;
};

export type SnackbarHideProps = {
  hide?: true;
  hideFadeOutDuration?: number;
};

export type SnackbarProps = SnackbarShowProps & SnackbarHideProps;

export const Snackbar: React.FC<SnackbarProps> = (props) => {
  const fadeAnim = useRef(new Animated.Value(0)).current;
  const fadeInAnimationRef = useRef<CompositeAnimation>();
  const fadeOutAnimationRef = useRef<CompositeAnimation>();
  const barrierFadeOutAnimationRef = useRef<CompositeAnimation>();
  const [visibleSnackbarProps, setVisibleSnackbarProps] = useState<SnackbarShowProps>();

  const animationStart = useCallback(
    (
      animationRef: React.MutableRefObject<CompositeAnimation | undefined>,
      config: Animated.TimingAnimationConfig,
      callback?: Animated.EndCallback,
    ) => {
      const animation = Animated.timing(fadeAnim, config);
      animationRef.current = animation;
      animation.start((result) => {
        animationRef.current = undefined;
        if (callback) {
          callback(result);
        }
      });
    },
    [fadeAnim],
  );

  const show = useCallback(() => {
    setVisibleSnackbarProps(props);

    const fadeInAnimationConfig = {
      toValue: 1,
      duration: props.fadeInDuration,
      useNativeDriver: true,
    };
    animationStart(fadeInAnimationRef, fadeInAnimationConfig, ({finished}) => {
      if (finished) {
        const fadeOutAnimationConfig = {
          toValue: 0,
          delay: props.autoHideDuration,
          duration: props.fadeOutDuration,
          useNativeDriver: true,
        };
        animationStart(fadeOutAnimationRef, fadeOutAnimationConfig, () => {
          if (!barrierFadeOutAnimationRef) {
            setVisibleSnackbarProps(undefined);
          }
        });
      } else {
        if (!barrierFadeOutAnimationRef) {
          setVisibleSnackbarProps(undefined);
        }
      }
    });
  }, [props, animationStart]);

  const forceFadeout = useCallback(
    (fadeOutDuration?: number, callback?: () => void) => {
      const barrierFadeOutAnimationConfig = {
        toValue: 0,
        duration: fadeOutDuration,
        useNativeDriver: true,
      };

      fadeInAnimationRef.current?.stop();
      fadeOutAnimationRef.current?.stop();

      animationStart(barrierFadeOutAnimationRef, barrierFadeOutAnimationConfig, () => {
        setVisibleSnackbarProps(undefined);
        callback?.();
      });
    },
    [animationStart],
  );

  React.useEffect(() => {
    if (props.hide) {
      forceFadeout(props.hideFadeOutDuration ?? props.forceFadeOutDuration);
      return;
    }
    if (!props.message) {
      return;
    }
    if (barrierFadeOutAnimationRef.current) {
      return;
    }
    if (fadeInAnimationRef.current || fadeOutAnimationRef.current) {
      forceFadeout(props.forceFadeOutDuration, show);
      return;
    }
    show();
  }, [props, show, forceFadeout]);

  const snackbarStyle = StyleSheet.flatten([styles.snackbar, props.style]);

  const animatedViewStyle = StyleSheet.flatten<ViewStyle>([styles.animatedContainer, props.animatedContainerStyle]);
  if (animatedViewStyle?.top === undefined && animatedViewStyle.bottom === undefined) {
    animatedViewStyle.bottom = 20;
  }

  return (
    <>
      {props.children}
      {visibleSnackbarProps && (
        <FullWindowOverlay>
          <Animated.View style={StyleSheet.flatten([{opacity: fadeAnim}, animatedViewStyle])}>
            <View style={snackbarStyle}>
              <View style={styles.messageContainer}>
                <Text style={styles.messageText}>{visibleSnackbarProps.message}</Text>
              </View>
              {visibleSnackbarProps.actionText && visibleSnackbarProps.actionHandler && (
                <View style={styles.actionContainer}>
                  <TouchableOpacity onPress={visibleSnackbarProps.actionHandler}>
                    <Text style={styles.actionText}>{visibleSnackbarProps.actionText}</Text>
                  </TouchableOpacity>
                </View>
              )}
            </View>
          </Animated.View>
        </FullWindowOverlay>
      )}
    </>
  );
};

Snackbar.defaultProps = {
  autoHideDuration: 4000,
  fadeInDuration: 1000,
  fadeOutDuration: 1000,
  forceFadeOutDuration: 300,
};

const styles = StyleSheet.create({
  animatedContainer: {
    flex: 1,
    position: 'absolute',
    minHeight: 48,
    paddingHorizontal: 10,
    width: '100%',
  },
  snackbar: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 5,
    backgroundColor: '#393939',
  },
  messageContainer: {
    flex: 1,
    paddingVertical: 20,
    paddingLeft: 20,
    paddingRight: 10,
  },
  actionContainer: {
    flex: 0,
    justifyContent: 'flex-end',
    alignItems: 'center',
    paddingVertical: 20,
    paddingLeft: 10,
    paddingRight: 20,
  },
  messageText: {
    color: 'white',
  },
  actionText: {
    color: '#85A3F1',
  },
});
