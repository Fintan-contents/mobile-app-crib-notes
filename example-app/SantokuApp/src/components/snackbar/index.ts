export * from './FullWindowOverlay';
export * from './Snackbar';
export * from './WithSnackbar';
