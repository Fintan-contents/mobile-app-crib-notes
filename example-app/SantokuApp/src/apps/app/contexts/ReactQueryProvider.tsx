import NetInfo from '@react-native-community/netinfo';
import React, {useEffect, useMemo} from 'react';
import {AppState, AppStateStatus, Platform} from 'react-native';
import {focusManager, onlineManager, QueryClient, QueryClientProvider} from 'react-query';

import {defaultMutationCache, defaultQueryCache} from '../services/defaultCache';
import {defaultOptions} from '../services/defaultOptions';

const onAppStateChange = (newAppState: AppStateStatus) => {
  if (Platform.OS !== 'web') {
    focusManager.setFocused(newAppState === 'active');
  }
};

export const ReactQueryProvider: React.FC<React.PropsWithChildren> = ({children}) => {
  const queryClient: QueryClient = useMemo(() => {
    return new QueryClient({
      queryCache: defaultQueryCache(queryClient),
      mutationCache: defaultMutationCache(queryClient),
      defaultOptions,
    });
  }, []);

  useEffect(() => {
    // アプリがバックグラウンドからアクティブに変化した際にrefetchできるようにする
    const subscription = AppState.addEventListener('change', onAppStateChange);
    return () => subscription.remove();
  }, []);

  useEffect(() => {
    // オフライン状態からオンライン状態に変化した際にrefetchできるようにする
    onlineManager.setEventListener(setOnline => {
      return NetInfo.addEventListener(state => {
        setOnline(state.isConnected ?? false);
      });
    });
  }, []);

  return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>;
};
