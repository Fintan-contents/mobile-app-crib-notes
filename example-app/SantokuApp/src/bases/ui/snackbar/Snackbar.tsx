import {m} from 'bases/message/Message';
import React, {useMemo, useState} from 'react';

import {
  SnackbarComponent,
  SnackbarHideProps as SnackbarComponentHideProps,
  SnackbarProps as SnackbarComponentProps,
  SnackbarShowProps as SnackbarComponentShowProps,
} from './SnackbarComponent';
import {FullWindowOverlay} from '../overlay/FullWindowOverlay';

type SnackbarShowProps = Omit<SnackbarComponentShowProps, 'message'>;
type SnackbarShowCloseButtonProps = Omit<SnackbarShowProps, 'actionText' | 'actionHandler' | 'actionTextStyle'>;
type SnackbarHideProps = Omit<SnackbarComponentHideProps, 'hide'>;

type SnackbarType = {
  /**
   * Show snackbar.
   *
   * @param message - Displayed message.
   * @param showProps - Snackbar props(SnackbarShowContextProps).
   */
  show: (message: string, showProps?: SnackbarShowProps) => void;

  /**
   * Show the snackbar with the close button.
   * The close button is placed to the right of the message.
   *
   * @param message - Displayed message.
   * @param showProps - Snackbar props(SnackbarShowCloseButtonContextProps).
   */
  showWithCloseButton: (message: string, showProps?: SnackbarShowCloseButtonProps) => void;

  /**
   * Hide snackbar.
   *
   * @param hideProps - Snackbar props(SnackbarHideContextProps).
   */
  hide: (hideProps?: SnackbarHideProps) => void;
  Component: typeof Component;
};

let Snackbar: SnackbarType = {
  show: () => {
    throw new Error('Snackbar.Component is required.');
  },
  showWithCloseButton: () => {
    throw new Error('Snackbar.Component is required.');
  },
  hide: () => {
    throw new Error('Snackbar.Component is required.');
  },
  Component,
};

function Component(props: {initialState?: SnackbarComponentShowProps}) {
  const [state, setState] = useState<SnackbarComponentProps>(
    props.initialState ?? {
      message: '',
    },
  );
  // const snackbarContext = useMemo<SnackbarContextType>(
  Snackbar = useMemo<SnackbarType>(
    () => ({
      ...Snackbar,
      show: (message: string, showProps?: SnackbarShowProps) => {
        setState({
          timestamp: Date.now(),
          ...showProps,
          message,
        });
      },
      showWithCloseButton: (message: string, showProps?: SnackbarShowCloseButtonProps) => {
        setState({
          timestamp: Date.now(),
          ...showProps,
          actionText: m('閉じる'),
          actionHandler: Snackbar.hide,
          message,
        });
      },
      hide: (hideProps?: SnackbarHideProps) => {
        setState(prevState => ({...prevState, ...hideProps, hide: true}));
      },
    }),
    [],
  );
  return (
    <>
      <FullWindowOverlay>
        <SnackbarComponent {...state} />
      </FullWindowOverlay>
    </>
  );
}

export {Snackbar};
