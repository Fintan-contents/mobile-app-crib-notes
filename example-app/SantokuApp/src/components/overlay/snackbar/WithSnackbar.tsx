import {m} from 'framework';
import {createUseContextAndProvider} from 'framework/utilities';
import React, {useMemo, useState} from 'react';

import {FullWindowOverlay} from '../FullWindowOverlay';
import {Snackbar, SnackbarHideProps, SnackbarProps, SnackbarShowProps} from './Snackbar';

type SnackbarShowContextProps = Omit<SnackbarShowProps, 'message'>;
type SnackbarShowCloseButtonContextProps = Omit<
  SnackbarShowContextProps,
  'actionText' | 'actionHandler' | 'actionTextStyle'
>;
type SnackbarHideContextProps = Omit<SnackbarHideProps, 'hide'>;

type SnackbarContextType = {
  /**
   * Show snackbar.
   *
   * @param message - Displayed message.
   * @param showProps - Snackbar props(SnackbarShowContextProps).
   */
  show: (message: string, showProps?: SnackbarShowContextProps) => void;

  /**
   * Show the snackbar with the close button.
   * The close button is placed to the right of the message.
   *
   * @param message - Displayed message.
   * @param showProps - Snackbar props(SnackbarShowCloseButtonContextProps).
   */
  showWithCloseButton: (message: string, showProps?: SnackbarShowCloseButtonContextProps) => void;

  /**
   * Hide snackbar.
   *
   * @param hideProps - Snackbar props(SnackbarHideContextProps).
   */
  hide: (hideProps?: SnackbarHideContextProps) => void;
};

const [useSnackbar, SnackbarContextProvider] = createUseContextAndProvider<SnackbarContextType>();

function WithSnackbar(props: {initialState?: SnackbarShowProps; children: React.ReactNode}) {
  const [state, setState] = useState<SnackbarProps>(
    props.initialState ?? {
      message: '',
    },
  );
  const snackbarContext = useMemo<SnackbarContextType>(
    () => ({
      show: (message: string, showProps?: SnackbarShowContextProps) => {
        setState({
          timestamp: Date.now(),
          ...showProps,
          message,
        });
      },
      showWithCloseButton: (message: string, showProps?: SnackbarShowCloseButtonContextProps) => {
        setState({
          timestamp: Date.now(),
          ...showProps,
          actionText: m('閉じる'),
          actionHandler: snackbarContext.hide,
          message,
        });
      },
      hide: (hideProps?: SnackbarHideContextProps) => {
        setState(prevState => ({...prevState, ...hideProps, hide: true}));
      },
    }),
    [],
  );
  return (
    <SnackbarContextProvider value={snackbarContext}>
      {props.children}
      <FullWindowOverlay>
        <Snackbar {...state} />
      </FullWindowOverlay>
    </SnackbarContextProvider>
  );
}

export {useSnackbar, WithSnackbar};
