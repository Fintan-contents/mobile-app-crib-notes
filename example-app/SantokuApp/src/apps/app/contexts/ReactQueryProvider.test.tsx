import {render, screen} from '@testing-library/react-native';
import {Snackbar} from 'bases/ui/snackbar/Snackbar';
import React from 'react';
import {Text} from 'react-native';
import {QueryClient, QueryClientProvider} from 'react-query';

import {ReactQueryProvider} from './ReactQueryProvider';

const Wrapper: React.FC<React.PropsWithChildren> = ({children}) => {
  const queryClient = new QueryClient();
  return (
    <>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
      <Snackbar.Component />
    </>
  );
};

describe('ReactQueryProvider', () => {
  test('子要素を正常にrenderできること', () => {
    const children = <Text testID="test">test</Text>;
    render(<ReactQueryProvider>{children}</ReactQueryProvider>, {wrapper: Wrapper});
    expect(screen.getByTestId('test')).not.toBeNull();
    expect(screen).toMatchSnapshot();
  });
});
