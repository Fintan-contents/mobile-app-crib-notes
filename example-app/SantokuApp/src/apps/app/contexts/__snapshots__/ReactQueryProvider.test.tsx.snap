// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactQueryProvider 子要素を正常にrenderできること 1`] = `
Array [
  <Text
    testID="test"
  >
    test
  </Text>,
  <View
    testID="FullWindowOverlay"
  />,
]
`;
